src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -134,13 +177,136 @@ private void SetSubFieldNodes(ExecutionContext context, ObjectExecutionNode pare
+        /// </summary>
+        protected virtual FieldType GetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
+        {
+            if (field.Name == schema.SchemaMetaFieldType.Name && schema.Query == parentType)
+            {
+                return schema.SchemaMetaFieldType;
+            }
+            if (field.Name == schema.TypeMetaFieldType.Name && schema.Query == parentType)
+            {
+                return schema.TypeMetaFieldType;
+            }
+            if (field.Name == schema.TypeNameMetaFieldType.Name)
+            {
+                return schema.TypeNameMetaFieldType;
+            }
+
+            if (parentType == null)
+            {
+                throw new ArgumentNullException(nameof(parentType), $"Schema is not configured correctly to fetch field '{field.Name}'. Are you missing a root type?");
+            }
+
+            return parentType.GetField(field.Name);
+        }
+
+        /// <inheritdoc/>
+        public virtual Fields GetSubFields(ExecutionContext context, ExecutionNode node)
+        {
+            return node.Field?.SelectionSet?.Selections?.Count > 0
+                ? CollectFieldsFrom(context, node.FieldDefinition.ResolvedType, node.Field.SelectionSet)
+                : null;
+        }
+
+        /// <summary>
+        /// Before execution, the selection set is converted to a grouped field set by calling CollectFields().
+        /// Each entry in the grouped field set is a list of fields that share a response key (the alias if defined,
+        /// otherwise the field name). This ensures all fields with the same response key included via referenced
+        /// fragments are executed at the same time.
+        /// <br/><br/>
+        /// <see href="http://spec.graphql.org/June2018/#sec-Field-Collection"/> and <see href="http://spec.graphql.org/June2018/#CollectFields()"/>
+        /// </summary>
+        /// <param name="context">The execution context.</param>
+        /// <param name="specificType">The graph type to compare the selection set against.</param>
+        /// <param name="selectionSet">The selection set from the document.</param>
+        /// <param name="fields">A list to append the collected list of fields to; if null, a new list will be created.</param>
+        /// <returns>A list of collected fields</returns>
+        protected virtual Fields CollectFieldsFrom(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Fields fields = null)
+        {
+            fields ??= new Fields();
+            List<string> visitedFragmentNames = null;
+            CollectFields(context, specificType.GetNamedType(), selectionSet, fields, ref visitedFragmentNames);
+            return fields;
+
+            void CollectFields(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Fields fields, ref List<string> visitedFragmentNames) //TODO: can be completely eliminated? see Fields.Add
```

Comment:
Moved from `Fields` (previously in `ExecutionHelper`).  I think it belongs here, as execution-related code, rather than in `ExecutionHelper` as a static method, or in `Fields`, which should be a simple collection class.